src/main/java/org/threadly/concurrent/PriorityScheduler.java

```diff
@@ -812,101 +731,204 @@ public int getCurrentRunningCount() {
-      synchronized (workersLock) {
-        if (shutdownFinishing || currentPoolSize > maxPoolSize) {
-          killWorker(worker);
-        } else {
-          // always add to the front so older workers are at the back
-          availableWorkers.addFirst(worker);
-          
-          if (waitingForWorker) {
-            workersLock.notify();
+    public TaskWrapper workerIdle(Worker worker) {
+      boolean interruptedChecked = false;
+      boolean queued = false;
+      try {
+        idle: while (true) {
+          int casPoolSize;
+          if (shutdownFinishing) {
+            currentPoolSize.decrementAndGet();
+            worker.stopIfRunning();
+            return null;
+          } else if ((casPoolSize = currentPoolSize.get()) > maxPoolSize) {
+            if (currentPoolSize.compareAndSet(casPoolSize, casPoolSize - 1)) {
+              worker.stopIfRunning();
+              return null;
+            }
+          } else {
+            // ready to poll task, we loop here to avoid pool state recheck logic unless we have to block
+            while (true) {
+              TaskWrapper nextTask = queueManager.getNextTask();
+              if (nextTask == null) {
+                if (queued) {
+                  // we can only park after we have queued, then checked again for a result
+                  LockSupport.park();
+                  continue idle;
+                } else {
+                  addWorkerToIdleChain(worker);
+                  queued = true;
+                }
+              } else {
+                /* TODO - right now this has a a deficiency where a recurring period task can cut in 
+                 * the queue line.  The condition would be as follows:
+                 * 
+                 * * Thread 1 gets task to run...task is behind execution schedule, likely due to large queue
+                 * * Thread 2 gets same task
+                 * * Thread 1 gets reference, executes, task execution completes
+                 * * Thread 2 now gets the reference, and execution check and time check pass fine
+                 * * End result is that task has executed twice (on expected schedule), the second 
+                 *     execution was unfair since it was done without respects to queue order and 
+                 *     other tasks which are also likely behind execution schedule in this example
+                 *     
+                 * This should be very rare, but is possible.  The only way I see to solve this right 
+                 * now is to introduce locking.
+                 */
+                // must get executeReference before time is checked
+                short executeReference = nextTask.getExecuteReference();
+                long taskDelay = nextTask.getScheduleDelay();
+                if (taskDelay > 0) {
+                  if (taskDelay == Long.MAX_VALUE) {
+                    // TODO - this will cause a spin if there is only one recurring task
```

Comment:
This is a MUCH more major issue....this will cause the CPU to spin if the pool only has recurring tasks, and all recurring tasks are currently execution, and there are idle threads trying to get additional work.

The condition here is that we now have taken a task that another thread also took and started to execute.  We don't want to park forever, because we really just got unlucky in getting this task.  For that reason a yield and retry works in the normal case.

Of course if the queue only contains executing recurring tasks we will keep retrying to get a new task.  In that condition we should be parked forever, but determining we are in this state is not trivial (suggestions on a cheap way we can detect it?).